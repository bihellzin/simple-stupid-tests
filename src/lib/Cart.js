import { find, remove } from 'lodash';
import Dinero from 'dinero.js';

const Money = Dinero;
Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;
export default class Cart {
  items = [];

  add(item) {
    const itemToFind = { product: item.product };
    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind);
    }
    this.items.push(item);
  }

  remove(product) {
    remove(this.items, { product });
  }

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc.add(Money({ amount: item.product.price * item.quantity }));
    }, Money({ amount: 0 }));
  }

  summary() {
    const total = this.getTotal().getAmount();
    const items = this.items;

    return {
      items,
      total,
    };
  }

  checkout() {
    const { total, items } = this.summary();

    this.items = [];
    return {
      items,
      total,
    };
  }
}
