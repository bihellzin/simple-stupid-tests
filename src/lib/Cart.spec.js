import Cart from './Cart';

describe('Cart', () => {
  let cart;
  let product = {
    title: 'aguinha geladinha',
    price: 150, // 1.50
  };

  let product2 = {
    title: 'pipoca bokus',
    price: 100, // 1.50
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe('getTotal()', () => {
    it('should return 0 when getTotal() is excecuted in a newly created cart', () => {
      expect(cart.getTotal()).toBe(0);
    });

    it('should multiply quantity and price and receive the total amount', () => {
      cart.add({
        product,
        quantity: 2, // 3.00
      });
      expect(cart.getTotal()).toBe(300);
    });

    it('should ensure no more than one product exists at a time', () => {
      cart.add({
        product,
        quantity: 2, // 3.00
      });
      cart.add({
        product,
        quantity: 1, // 1.50
      });

      expect(cart.getTotal()).toBe(150);
    });

    it('should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2, // 3.00
      });

      cart.add({
        product: product2,
        quantity: 1, // 1.00
      });

      expect(cart.getTotal()).toBe(400);

      cart.remove(product);
      expect(cart.getTotal()).toBe(100);
    });
  });

  describe('checkout()', () => {
    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 8,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should reset the cart when checkout() is called', () => {
      cart.add({
        product: product2,
        quantity: 3,
      });

      cart.checkout();

      expect(cart.getTotal()).toEqual(0);
    });
  });

  describe('summary()', () => {
    it('should return an object with the total and the list of items when summary() is called', () => {
      cart.add({
        product,
        quantity: 9,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotal()).toBeGreaterThan(0);
    });
  });
});
