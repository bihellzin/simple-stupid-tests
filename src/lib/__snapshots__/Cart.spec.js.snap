// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
{
  "items": [
    {
      "product": {
        "price": 150,
        "title": "aguinha geladinha",
      },
      "quantity": 8,
    },
    {
      "product": {
        "price": 100,
        "title": "pipoca bokus",
      },
      "quantity": 3,
    },
  ],
  "total": 1500,
}
`;

exports[`Cart summary() should return an object with the total and the list of items when summary() is called 1`] = `
{
  "items": [
    {
      "product": {
        "price": 150,
        "title": "aguinha geladinha",
      },
      "quantity": 9,
    },
    {
      "product": {
        "price": 100,
        "title": "pipoca bokus",
      },
      "quantity": 1,
    },
  ],
  "total": 1450,
}
`;
